src/opensynth/models/faraday/gaussian_mixture/gmm_model.py

```diff
@@ -1,194 +1,421 @@
-                random samples.
+        self.automatic_optimization = False
+        self.convergence_tolerance = convergence_tolerance
 
-        Attention:
-            This method does not automatically perform batching. If you need to
-                sample many
-            datapoints, call this method multiple times.
-        """
-        # First, we sample counts for each
-
-        # Normalise component probabilities after converting to float64
-        # This is necessary because numpy casts the probabilities to float64
-        # and we want to avoid precision issues resulting from this conversion
-        # See https://github.com/numpy/numpy/issues/8317.
-        _probs = self.component_probs.numpy().astype("float64")
-        component_counts = np.random.multinomial(
-            num_datapoints, _probs / _probs.sum()
+        # GMM params to sync across processes
+        self.weight_metric = gmm_metrics.WeightsMetric(self.num_components)
+        self.mean_metric = gmm_metrics.MeansMetric(
+            self.num_components, self.num_features
+        )
+        self.precision_cholesky_metric = gmm_metrics.PrecisionCholeskyMetric(
+            self.num_components, self.num_features
         )
+        self.covariance_metric = gmm_metrics.CovarianceMetric(
+            self.num_components, self.num_features
+        )
+        self.nll = gmm_metrics.NLLMetric()
 
-        # Then, we generate datapoints for each components
-        result = []
-        for i, count in enumerate(component_counts):
-            mean = self.means[i]
-            samples = torch.randn(
-                count.item(),
-                mean.size(0),
-                dtype=mean.dtype,
-                device=mean.device,
-            )
+        self.sync_on_batch = sync_on_batch
 
-            cholesky_precisions = self.precisions_cholesky[i]
-            # For complex covariance types, invert the
-            num_features = cholesky_precisions.size(-1)
-            target = torch.eye(
-                num_features,
-                dtype=cholesky_precisions.dtype,
-                device=cholesky_precisions.device,
-            )
-            chol_covariance = torch.linalg.solve_triangular(
-                cholesky_precisions, target, upper=True
-            ).t()
+    def configure_optimizers(self) -> None:
+        return None
 
-            scale = chol_covariance.matmul(samples.unsqueeze(-1)).squeeze(-1)
+    def on_train_epoch_start(self) -> None:
+        # At the start of epoch, reset metrics
+        self.mean_metric.reset()
+        self.weight_metric.reset()
+        self.precision_cholesky_metric.reset()
+        self.covariance_metric.reset()
+        self.nll.reset()
 
-            sample = mean + scale
+    def training_step(self, batch) -> None:
+        # Encode the batch
+        encoded_batch = gmm_utils.encode_data(batch, self.vae_module)
 
-            result.append(sample)
+        if "weights" in list(batch.keys()):
```

Comment:
I prefer we let end users define the name of the `sample_weight` column. This column might be named differently in different datasets and best we don't presume what the name of the column is.